File: ddz_admin/app/admin/task_def.rb
ActiveAdmin.register TaskDef do


  permit_params do
    permitted = []
    permitted << :taskId
    permitted << :taskName
    permitted << :taskDesc
    permitted << :taskType
    permitted << :taskIcon
    permitted << :taskBonusDesc
    permitted << :taskTrigger
    permitted << :taskProcessor
    permitted << :enabled
    permitted << :sortIndex
    permitted << :progress
    permitted << :progressDesc
    permitted << :taskData_hash
    permitted
  end

  index do
    selectable_column
    id_column
    column :taskId
    column :taskName
    column :taskDesc
    column :taskType
    column :taskIcon
    column :taskBonusDesc
    column :taskTrigger
    column :taskProcessor
    column :enabled
    column :sortIndex
    column :progress
    column :progressDesc
    column :taskData_hash
    column :created_at
    column :updated_at
    actions
  end

  filter :taskId
  filter :taskName
  filter :taskDesc
  filter :taskType

  form do |f|
    f.inputs "TaskDef Details" do
      f.input :taskId
      f.input :taskName
      f.input :taskDesc
      f.input :taskType
      f.input :taskIcon
      f.input :taskBonusDesc
      f.input :taskTrigger
      f.input :taskProcessor
      f.input :enabled
      f.input :sortIndex
      f.input :progress
      f.input :progressDesc
      f.input :taskData_hash, as: :text
    end

    f.actions
  end

end

File: login-server/app/domain/taskDef.js
/**
 * Created by edwardzhou on 14/12/4.
 */


var mongoose = require('mongoose-q')();
var Schema = mongoose.Schema;
var uuid = require('node-uuid');
var utils = require('../util/utils');

/**
 * 任务定义
 * @type {Mongoose.Schema}
 */
var taskDefSchema = new mongoose.Schema({
  taskId: {type: String, required: "{task id} is required!"},
  taskName: {type: String, required: "{task name} is required!"},
  taskDesc: {type: String, required: "{task description} is required!"},
  taskType: {type: String, required: true},
  taskIcon: {type: String},
  taskBonusDesc: {type: String},
  sortIndex: {type: Number, default: 255},
  enabled: {type: Boolean, default: true},
  progress: {type: Number, default: 0},
  progressDesc: String,
  taskProcessor: String,
  taskTrigger: String,
  taskData: {type: Schema.Types.Mixed, default: {_placeholder:0}},
  created_at: {type: Date, default: Date.now},
  updated_at: {type: Date, default: Date.now}
}, {
  collection: 'task_defs'
});


var __toParams = function(model, excludeAttrs) {
  var transObj = {
    userId: 0,
    taskId: model.taskId,
    taskName: model.taskName,
    taskDesc: model.taskDesc,
    taskIcon: model.taskIcon,
    taskType: model.taskType,
    taskBonusDesc: model.taskBonusDesc,
    progress: model.progress,
    progressDesc: model.progressDesc
  };

  if (!!excludeAttrs) {
    for (var index=0; index<excludeAttrs.length; index++) {
      delete transObj[excludeAttrs[index]];
    }
  }

  return transObj;
};

taskDefSchema.statics.toParams = __toParams;

taskDefSchema.methods.toParams = function(excludeAttrs) {
  return __toParams(this, excludeAttrs);
};


var TaskDef = mongoose.model('TaskDef', taskDefSchema);

module.exports = TaskDef;